### PR TITLE
[DO NOT SQUASH] Add --miopen-clean-math, integrate integer range analysis

### PR DESCRIPTION
**Pull request by krzysz00**
- Move the canonicalize + cse that used to be at the end of
--miopen-sugar-to-loops to --miopen-clean-math
- Make --miopen-clean-math include using integer range analysis ho
  rewrite away things that are known to be constant (mainly the "is
  negative" tests emitted by the implementation of mod in affine map
  lowerings)
- Also use --unsigned-when-equivalent on the results of that analysis
  to clear up stray and unneeded signed divisions and remainders
- Define the maximum values of miopen.workgroup_id and
  miopen.workitem_id to be based on the grid_size and block_size
  arguments to their containing functions if they exist and to be
  2^31 - 1 otherwise

Note that I had to copy in what is, upstream, a test pass, and that
will hopefully go away in the future.

The patch to the truncation logic is successfully landed upstream at https://reviews.llvm.org/D130501

### CODE REVIEW

**Comment by krzysz00** [comment]
To give an example, the initial commputations right before the writeback loop used to be
```
      memref.store %1388, %arg4[%c63] : memref<64xf32, 5>
      %1389 = arith.cmpi slt, %0, %c0 : index
      %1390 = arith.subi %c-1, %0 : index
      %1391 = arith.select %1389, %1390, %0 : index
      %1392 = arith.divsi %1391, %c1568 : index
      %1393 = arith.subi %c-1, %1392 : index
      %1394 = arith.select %1389, %1393, %1392 : index
      %1395 = arith.remsi %0, %c1568 : index
      %1396 = arith.cmpi slt, %1395, %c0 : index
      %1397 = arith.addi %1395, %c1568 : index
      %1398 = arith.select %1396, %1397, %1395 : index
      %1399 = arith.cmpi slt, %1398, %c0 : index
      %1400 = arith.subi %c-1, %1398 : index
      %1401 = arith.select %1399, %1400, %1398 : index
      %1402 = arith.divsi %1401, %c8 : index
      %1403 = arith.subi %c-1, %1402 : index
      %1404 = arith.select %1399, %1403, %1402 : index
      %1405 = arith.remsi %0, %c8 : index
      %1406 = arith.cmpi slt, %1405, %c0 : index
      %1407 = arith.addi %1405, %c8 : index
      %1408 = arith.select %1406, %1407, %1405 : index
      %1409 = arith.cmpi slt, %1, %c0 : index
      %1410 = arith.subi %c-1, %1 : index
      %1411 = arith.select %1409, %1410, %1 : index
      %1412 = arith.divsi %1411, %c64 : index
      %1413 = arith.subi %c-1, %1412 : index
      %1414 = arith.select %1409, %1413, %1412 : index
      %1415 = arith.remsi %1, %c64 : index
      %1416 = arith.cmpi slt, %1415, %c0 : index
      %1417 = arith.addi %1415, %c64 : index
      %1418 = arith.select %1416, %1417, %1415 : index
      %1419 = arith.cmpi slt, %1418, %c0 : index
      %1420 = arith.subi %c-1, %1418 : index
      %1421 = arith.select %1419, %1420, %1418 : index
      %1422 = arith.divsi %1421, %c32 : index
      %1423 = arith.subi %c-1, %1422 : index
      %1424 = arith.select %1419, %1423, %1422 : index
      %1425 = arith.remsi %1, %c32 : index
      %1426 = arith.cmpi slt, %1425, %c0 : index
      %1427 = arith.addi %1425, %c32 : index
      %1428 = arith.select %1426, %1427, %1425 : index
      %1429 = arith.cmpi slt, %1428, %c0 : index
      %1430 = arith.subi %c-1, %1428 : index
      %1431 = arith.select %1429, %1430, %1428 : index
      %1432 = arith.divsi %1431, %c4 : index
      %1433 = arith.subi %c-1, %1432 : index
      %1434 = arith.select %1429, %1433, %1432 : index
      %1435 = arith.remsi %1, %c4 : index
      %1436 = arith.cmpi slt, %1435, %c0 : index
      %1437 = arith.addi %1435, %c4 : index
      %1438 = arith.select %1436, %1437, %1435 : index
      %1439 = arith.cmpi slt, %1414, %c0 : index
      %1440 = arith.subi %c-1, %1414 : index
      %1441 = arith.select %1439, %1440, %1414 : index
      %1442 = arith.divsi %1441, %c2 : index
      %1443 = arith.subi %c-1, %1442 : index
      %1444 = arith.select %1439, %1443, %1442 : index
      %1445 = arith.remsi %1414, %c2 : index
      %1446 = arith.cmpi slt, %1445, %c0 : index
      %1447 = arith.addi %1445, %c2 : index
      %1448 = arith.select %1446, %1447, %1445 : index
      %1449 = arith.muli %1408, %c128 : index
      %1450 = arith.muli %1444, %c64 : index
      %1451 = arith.addi %1449, %1450 : index
      %1452 = arith.muli %1424, %c4 : index
      %1453 = arith.addi %1451, %1452 : index
      %1454 = arith.addi %1453, %1438 : index
      %1455 = arith.muli %1404, %c128 : index
      %1456 = arith.muli %1448, %c64 : index
      %1457 = arith.addi %1455, %1456 : index
      %1458 = arith.muli %1434, %c4 : index
      %1459 = arith.addi %1457, %1458 : index
      %1460 = arith.cmpi slt, %1459, %c0 : index
      %1461 = arith.subi %c-1, %1459 : index
      %1462 = arith.select %1460, %1461, %1459 : index
      %1463 = arith.divsi %1462, %c196 : index
      %1464 = arith.subi %c-1, %1463 : index
      %1465 = arith.select %1460, %1464, %1463 : index
      %1466 = arith.remsi %1459, %c196 : index
      %1467 = arith.cmpi slt, %1466, %c0 : index
      %1468 = arith.addi %1466, %c196 : index
      %1469 = arith.select %1467, %1468, %1466 : index
      %1470 = arith.cmpi slt, %1469, %c0 : index
      %1471 = arith.subi %c-1, %1469 : index
      %1472 = arith.select %1470, %1471, %1469 : index
      %1473 = arith.divsi %1472, %c14 : index
      %1474 = arith.subi %c-1, %1473 : index
      %1475 = arith.select %1470, %1474, %1473 : index
      %1476 = arith.remsi %1459, %c14 : index
      %1477 = arith.cmpi slt, %1476, %c0 : index
      %1478 = arith.addi %1476, %c14 : index
      %1479 = arith.select %1477, %1478, %1476 : index
      %1480 = memref.load %arg4[%c0] : memref<64xf32, 5>
...
      amdgpu.raw_buffer_store {boundsCheck = true} %1487 -> %arg2[%1488, %1489, %1490, %1491, %1492] : vector<4xf32> -> memref<128x1x1024x14x14xf32>, i32, i32, i32, i32, i32
```

and is now replaced with
```
      memref.store %1350, %arg4[%c63] : memref<64xf32, 5>
      %1351 = arith.remui %0, %c1568 : index
      %1352 = arith.divui %1351, %c8 : index
      %1353 = arith.remui %0, %c8 : index
      %1354 = arith.divui %1, %c64 : index
      %1355 = arith.remui %1, %c64 : index
      %1356 = arith.divui %1355, %c32 : index
      %1357 = arith.remui %1, %c32 : index
      %1358 = arith.divui %1357, %c4 : index
      %1359 = arith.remui %1, %c4 : index
      %1360 = arith.divui %1354, %c2 : index
      %1361 = arith.remui %1354, %c2 : index
      %1362 = arith.muli %1353, %c128 : index
      %1363 = arith.muli %1360, %c64 : index
      %1364 = arith.addi %1362, %1363 : index
      %1365 = arith.muli %1356, %c4 : index
      %1366 = arith.addi %1364, %1365 : index
      %1367 = arith.addi %1366, %1359 : index
      %1368 = arith.muli %1352, %c128 : index
      %1369 = arith.muli %1361, %c64 : index
      %1369 = arith.muli %1361, %c64 : index
      %1370 = arith.addi %1368, %1369 : index
      %1371 = arith.muli %1358, %c4 : index
      %1372 = arith.addi %1370, %1371 : index
      %1373 = arith.divui %1372, %c196 : index
      %1374 = arith.remui %1372, %c196 : index
      %1375 = arith.divui %1374, %c14 : index
      %1376 = arith.remui %1372, %c14 : index
      %1377 = memref.load %arg4[%c0] : memref<64xf32, 5>
...
      amdgpu.raw_buffer_store {boundsCheck = true} %1384 -> %arg2[%1385, %c0_i32, %1386, %1387, %1388] : vector<4xf32> -> memref<128x1x1024x14x14xf32>, i32, i32, i32, i32, i32
```

This is on `-p -x2`

(and here's an extract from the ISA diff, with register numbers replaced by `?` to make this diff only a bit noisy)
```
-       s_ashr_i32 s?, s?, 31
+       v_cmp_eq_u32_e32 vcc, 0, v?
        s_waitcnt lgkmcnt(2)
        v_mfma_f32_32x32x1f32 a?, v?, v?, a? cbsz:1
-       s_xor_b32 s?, s?, s?
-       s_mul_hi_i32 s?, s?, 0x5397829d
-       s_lshr_b32 s?, s?, 31
-       s_ashr_i32 s?, s?, 9
-       s_add_i32 s?, s?, s?
-       s_xor_b32 s?, s?, s?
-       s_mul_hi_i32 s?, s?, 0x5397829d
-       s_lshr_b32 s?, s?, 31
-       s_ashr_i32 s?, s?, 9
-       s_add_i32 s?, s?, s?
-       s_mulk_i32 s?, 0x620
-       s_sub_i32 s?, s?, s?
-       s_add_i32 s?, s?, 0x620
-       s_cmp_lt_i32 s?, 0
-       s_cselect_b32 s?, s?, s?
-       v_mfma_f32_32x32x1f32 a?, v?, v?, a? cbsz:1
-       s_lshr_b32 s?, s?, 29
-       s_add_i32 s?, s?, s?
-       s_and_b32 s?, s?, -8
-       s_sub_i32 s?, s?, s?
-       s_lshr_b32 s?, s?, 31
-       s_lshr_b32 s?, s?, 3
-       s_lshl_b32 s?, s?, 7
-       v_cmp_eq_u32_e32 vcc, 0, v?
-       s_xor_b32 s?, s?, s?
-       s_add_i32 s?, s?, 0x400
-       s_cmp_lt_i32 s?, 0
-       v_cmp_gt_i32_e64 s?, 0, v?
        s_mov_b32 s?, 0x5397829d
        v_cmp_eq_u32_e64 s?, 0, v?
-       s_mul_i32 s?, s?, 0xc4000
+       s_movk_i32 s?, 0x310
```
and
```
-       v_cndmask_b32_e64 v?, v?, v?, s?
-       v_and_b32_e32 v?, 31, v?
+       v_or_b32_e32 v?, s?, v?
        v_cndmask_b32_e32 v?, v?, v?, vcc
        v_cndmask_b32_e32 v?, v?, v?, vcc
-       v_or_b32_e32 v?, 32, v?
-       v_cmp_gt_i32_e64 s?, 0, v?
-       v_cndmask_b32_e64 v?, v?, v?, s?
-       v_and_b32_e32 v?, 3, v?
-       v_or_b32_e32 v?, 4, v?
-       v_cmp_gt_i32_e64 s?, 0, v?
-       v_bfe_u32 v?, v?, 6, 1
-       v_cndmask_b32_e64 v?, v?, v?, s?
-       s_cselect_b32 s?, s?, s?
-       v_lshlrev_b32_e32 v?, 6, v?
+       v_mul_hi_u32 v?, v?, s?
+       s_lshl_b32 s?, s?, 7
+       s_and_b32 s?, s?, 0x380
```

Perf numbers coming soon

**Comment by krzysz00** [comment]
Don't merge, there's a bug _somewhere_

**Comment by krzysz00** [comment]
To just down my notes on the bug hunt
1. Looks like a bug of some flavor in how the range analysis propagates detecting potentially loop-variant results through stuff like index_cast - this could be a more general issue upstream, will poke at it
2. I could probably do with making loop variant value detection a bit more optimistic (ex. having it keep shared lower bounds if it finds them)

**Comment by krzysz00** [comment]
Update: fixed bug, was an issue with inferring truncation

**Comment by krzysz00** [comment]
Perf impact: generally a minor (2-ish%) increase, with some weird outliers that seem to be a result of frequency scaling... 
[range-analysis-perf.zip](https://github.com/ROCmSoftwarePlatform/llvm-project-mlir/files/9184295/range-analysis-perf.zip)

**Comment by krzysz00** [comment]
Update sanity, I forgot

**Comment by krzysz00** [comment]
And I've updated the sanity tests - this should all be mergeable now

**Comment by krzysz00** [comment]
Note: the bug fix around truncation has been upstreamed, and the commit on this branch reflects exactly what landed upstream

File: external/llvm-project/mlir/lib/Dialect/Arithmetic/IR/InferIntRangeInterfaceImpls.cpp [hunk 1]
```diff
@@ -503,10 +503,37 @@ void arith::ExtSIOp::inferResultRanges(ArrayRef<ConstantIntRanges> argRanges,
 static ConstantIntRanges truncIRange(const ConstantIntRanges &range,
                                      Type destType) {
   unsigned destWidth = ConstantIntRanges::getStorageBitwidth(destType);
-  APInt umin = range.umin().trunc(destWidth);
-  APInt umax = range.umax().trunc(destWidth);
-  APInt smin = range.smin().trunc(destWidth);
-  APInt smax = range.smax().trunc(destWidth);
+  // If you truncate the first four bytes in [0xaaaabbbb, 0xccccbbbb],
```
Comment (jerryyin) [inline external/llvm-project/mlir/lib/Dialect/Arithmetic/IR/InferIntRangeInterfaceImpls.cpp:506]: Add a reference in this PR about the upstream review link?

**Review by jerryyin — APPROVED** [review]
LGTM

**Pull request merged**